src/workerd/jsg/modules-new.c++

```diff
@@ -378,41 +375,62 @@ class IsolateModuleRegistry final {
       auto module = entry.key.getHandle(js);
       auto status = module->GetStatus();
 
-      if (status == v8::Module::kEvaluated) {
-        return module->GetModuleNamespace().As<v8::Object>();
+      // If status is kErrored, that means a prior attempt to evaluate the module
+      // failed. We simply propagate the same error here.
+      if (status == v8::Module::kErrored) {
+        js.throwException(JsValue(module->GetException()));
       }
 
-      if (status == v8::Module::kErrored) {
-        js.throwException(js.v8Ref(module->GetException()));
+      // TODO(new-module-registry): Circular dependencies should be fine
+      // when we are talking strictly about CJS/Node.js style modules.
+      // For ESM, it becomes more problematic because v8 will not allow
+      // us to grab the default export while the module is still evaluating.
+
+      if (entry.module.isEsm() && status == v8::Module::kEvaluating) {
+        JSG_FAIL_REQUIRE(Error, "Circular dependency when resolving module: ", specifier);
       }
```

Comment:
I guess this can be simplified to JSG_REQUIRE(entry.module.isEsm() ...)

Reply:
Yes but I find this form more readable, to be honest. To use `JSG_REQUIRE(...)` would involve inverting the check here and honestly this just *reads* better to me than the alternative. I'd rather keep it as is.